File: axi/src/axi/sync/phrase.py
"""The twelve words: BIP-39 encoding, decoding, and a checksum that bites.

The phrase is the ONLY thing that can restore a user's ability to read their
own synced data. There is no escrow, no copy on the VPS, no reset link. That
makes two properties non-negotiable here:

  * DETERMINISM — the same twelve words must produce the same entropy on every
    device, forever, across both this implementation and the Dart one.
  * REJECTING A TYPO LOUDLY — a mistyped word must fail before it can derive
    anything. Deriving *some* key from a wrong phrase gives the user an app
    that silently opens nothing, with no error to act on.

Standard BIP-39, 128 bits of entropy: 128 + a 4-bit SHA-256 checksum = 132
bits = 12 words of 11 bits. No passphrase — see the module docstring in
`tests/test_sync_vectors.py` for why a second secret is a liability, not a
feature.
"""

from __future__ import annotations

import hashlib
import unicodedata
from pathlib import Path

#: The vendored official BIP-39 English wordlist. Vendored, not imported from a
#: package: these 2048 words in this exact order ARE the encoding. A dependency
#: that reordered or renormalised them would invalidate every phrase ever
#: written down, with no error anyone could read. Its content hash is asserted
#: in the tests.
WORDLIST_PATH = Path(__file__).with_name("wordlist") / "english.txt"

WORDS: tuple[str, ...] = tuple(
    WORDLIST_PATH.read_text(encoding="utf-8").split()
)

_INDEX: dict[str, int] = {w: i for i, w in enumerate(WORDS)}

ENTROPY_BYTES = 16
WORD_COUNT = 12


class InvalidPhrase(ValueError):
    """The phrase is not a valid LifeOS recovery phrase.

    Raised for a wrong word count, a word outside the list, or a failed
    checksum. Deliberately ONE exception type with a specific message rather
    than a family: the caller's job is to tell the user "that phrase is not
    right, check it and try again", and never to attempt a derivation anyway.
    """


def normalise(text: str) -> str:
    """Collapse what a human types into the canonical form.

    Case, leading/trailing padding and repeated whitespace are typing, not a
    different phrase. NFKD matches BIP-39's own normalisation, so a phrase
    pasted from a device with different Unicode composition still decodes —
    the Dart side normalises identically or the two drift apart.
    """
    return " ".join(unicodedata.normalize("NFKD", text).lower().split())


def encode(entropy: bytes) -> str:
    """16 bytes of entropy -> the twelve words."""
    if len(entropy) != ENTROPY_BYTES:
        raise ValueError(
            f"LifeOS phrases are {ENTROPY_BYTES * 8}-bit; got {len(entropy) * 8}"
        )

    checksum = hashlib.sha256(entropy).digest()[0] >> 4  # top 4 bits
    bits = int.from_bytes(entropy, "big") << 4 | checksum

    words = [
        WORDS[(bits >> (11 * (WORD_COUNT - 1 - i))) & 0x7FF]
        for i in range(WORD_COUNT)
    ]
    return " ".join(words)


def decode(mnemonic: str) -> bytes:
    """The twelve words -> 16 bytes of entropy, or raise [InvalidPhrase].

    The checksum is verified BEFORE the entropy is returned, so no caller can
    accidentally derive a key from a phrase with a typo in it. That ordering is
    the whole point of this function.
    """
    words = normalise(mnemonic).split()

    if len(words) != WORD_COUNT:
        raise InvalidPhrase(
            f"a LifeOS recovery phrase has {WORD_COUNT} words; got {len(words)}"
        )

    bits = 0
    for word in words:
        index = _INDEX.get(word)
        if index is None:
            raise InvalidPhrase(f"'{word}' is not a recovery-phrase word")
        bits = bits << 11 | index

    checksum = bits & 0xF
    entropy = (bits >> 4).to_bytes(ENTROPY_BYTES, "big")

    if hashlib.sha256(entropy).digest()[0] >> 4 != checksum:
        # Every word was real and the length was right — only the checksum
        # says otherwise. This is the case a "are the words valid?" check
        # would wave through, and it is the common one: a single wrong word.
        raise InvalidPhrase(
            "that recovery phrase is not valid — one of the words is wrong"
        )

    return entropy

File: mobile/lib/core/sync/phrase.dart
// The twelve words, Dart side. Byte-identical to `axi/src/axi/sync/phrase.py`.
//
// Every decision here mirrors the Python module deliberately, because the two
// are one format with two implementations. Read that file's docstring for the
// reasoning; this one only notes where Dart forced a different shape.
//
// The phrase is the ONLY thing that restores a user's ability to read their own
// synced data — no escrow, no copy on the VPS, no reset link. So: deterministic
// encoding, and a checksum that rejects a typo BEFORE anything derives a key
// from it. Deriving *some* key from a wrong phrase gives the user an app that
// silently opens nothing.
import 'dart:convert';

import 'package:crypto/crypto.dart' show sha256;

import 'bip39_wordlist.dart';

/// 128 bits of entropy + a 4-bit SHA-256 checksum = 132 bits = 12 × 11 bits.
const int kEntropyBytes = 16;
const int kWordCount = 12;

/// Thrown for a wrong word count, a word outside the list, or a bad checksum.
///
/// One exception type on purpose, matching Python's `InvalidPhrase`: the
/// caller's job is to say "that phrase is not right, check it" and never to
/// attempt a derivation anyway.
class InvalidPhrase implements Exception {
  const InvalidPhrase(this.message);
  final String message;

  @override
  String toString() => 'InvalidPhrase: $message';
}

Map<String, int>? _indexCache;
Map<String, int> get _index =>
    _indexCache ??= {for (var i = 0; i < kBip39English.length; i++) kBip39English[i]: i};

/// Collapse what a human types into the canonical form.
///
/// Case, padding and repeated whitespace are typing, not a different phrase.
/// Python applies Unicode NFKD here; the English wordlist is pure ASCII, so for
/// this list lowercasing and whitespace folding produce the identical result —
/// and the shared vectors prove it rather than leaving it to argument. A
/// non-ASCII wordlist would need real NFKD on both sides.
String normalisePhrase(String text) =>
    text.toLowerCase().split(RegExp(r'\s+')).where((w) => w.isNotEmpty).join(' ');

/// [normalisePhrase], plus forgiveness for what a KEYBOARD adds.
///
/// Separate from `normalisePhrase` on purpose. That function's exact behaviour
/// is pinned by the shared Python vectors and defines what a phrase IS; this
/// one defines what we accept from a human typing on glass, and only ever runs
/// at the input boundary.
///
/// It exists because confirmation was rejecting words that were RIGHT. Gboard
/// turns a double space into ". ", people separate lists with commas, and a
/// wrapped line can leave a soft hyphen (U+00AD) the user cannot see. Telling
/// someone their correct phrase is wrong is worse than a crash: a crash gets
/// reported, this gets believed — they retype, fail again, and conclude they
/// wrote the words down wrong.
///
/// Punctuation becomes a SPACE rather than being deleted, so "abandon,ability"
/// reads as two words instead of the single invalid "abandonability".
String sanitiseTypedPhrase(String text) =>
    normalisePhrase(text.replaceAll(RegExp('[^A-Za-z\\s]'), ' '));

/// 16 bytes of entropy -> the twelve words.
String encodePhrase(List<int> entropy) {
  if (entropy.length != kEntropyBytes) {
    throw ArgumentError(
      'LifeOS phrases are ${kEntropyBytes * 8}-bit; got ${entropy.length * 8}',
    );
  }

  final checksum = sha256.convert(entropy).bytes[0] >> 4;

  // BigInt, not int: 132 bits does not fit in Dart's 64-bit int on native, and
  // on the web `int` is a double with 53 bits of precision. Using BigInt makes
  // this identical on every platform LifeOS ships to.
  var bits = BigInt.zero;
  for (final b in entropy) {
    bits = (bits << 8) | BigInt.from(b);
  }
  bits = (bits << 4) | BigInt.from(checksum);

  final mask = BigInt.from(0x7FF);
  final words = <String>[];
  for (var i = 0; i < kWordCount; i++) {
    final shift = 11 * (kWordCount - 1 - i);
    words.add(kBip39English[((bits >> shift) & mask).toInt()]);
  }
  return words.join(' ');
}

/// The twelve words -> 16 bytes of entropy, or throw [InvalidPhrase].
///
/// The checksum is verified BEFORE the entropy is returned, so no caller can
/// derive a key from a phrase with a typo in it. That ordering is the function.
List<int> decodePhrase(String mnemonic) {
  final words = normalisePhrase(mnemonic).split(' ').where((w) => w.isNotEmpty).toList();

  if (words.length != kWordCount) {
    throw InvalidPhrase(
      'a LifeOS recovery phrase has $kWordCount words; got ${words.length}',
    );
  }

  var bits = BigInt.zero;
  for (final word in words) {
    final i = _index[word];
    if (i == null) {
      throw InvalidPhrase("'$word' is not a recovery-phrase word");
    }
    bits = (bits << 11) | BigInt.from(i);
  }

  final checksum = (bits & BigInt.from(0xF)).toInt();
  final entropyBits = bits >> 4;

  final entropy = List<int>.filled(kEntropyBytes, 0);
  var remaining = entropyBits;
  for (var i = kEntropyBytes - 1; i >= 0; i--) {
    entropy[i] = (remaining & BigInt.from(0xFF)).toInt();
    remaining = remaining >> 8;
  }

  if (sha256.convert(entropy).bytes[0] >> 4 != checksum) {
    // Every word real, the length right, only the checksum wrong: the common
    // failure, and the one a word-membership check would wave through.
    throw const InvalidPhrase(
      'that recovery phrase is not valid — one of the words is wrong',
    );
  }

  return entropy;
}

/// Convenience for callers that hold text rather than bytes.
String utf8Normalise(String text) => utf8.decode(utf8.encode(normalisePhrase(text)));
